### PR TITLE
Deprecate use of global PRNG in strategies

### PR DESCRIPTION
**Pull request by Zac-HD**
Fixes https://github.com/HypothesisWorks/hypothesis/issues/3810.

### CODE REVIEW

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py [hunk 1]
```diff
@@ -2134,9 +2134,20 @@ def draw(self, strategy: SearchStrategy[Ex], label: Any = None) -> Ex:
...
+        if _global_rand_state != random.getstate():
+            # raise InvalidDefinition
+            note_deprecation(
+                "Do not use the `random` module inside strategies; instead "
+                "consider  `st.randoms()`, `st.sampled_from()`, etc.  "
+                f"{desc}from {strategy!r}",
+                since="RELEASEDAY",
+                has_codemod=False,
+                stacklevel=1,
+            )
```
Comment (jobh) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:2150]: Maybe consider factoring this out as

```
with checked_random_state():
    result = self.conjecture_data.draw(strategy, observe_as=f"generate:{desc}")
    ...
```
?

That would avoid the duplication w/control.py, plus improve readability in these central functions

**Review by jobh — APPROVED** [review]
Looks good!

An idea: What if we check more randomness sources, but for performance reasons only occasionally? For example every nth draw, or random spot checks with p<0.1?

I'm saying this because I'd typically reach to `np.random` instead of `math.random` myself.

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py (hunk 1 above)
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/strategies/_internal/core.py:2150]: Yeah, I like it - and we can check a "drew from `st.random_module()` flag there too.

**Comment by Zac-HD** [comment]
> An idea: What if we check more randomness sources, but for performance reasons only occasionally? For example every nth draw, or random spot checks with p<0.1?  I'm saying this because I'd typically reach to `np.random` instead of `math.random` myself.

Unfortunately Hypothesis will then complain about `Flaky` errors when it goes to shrink such things.

Maybe we check that less frequently?  Or only if you set a flag?  Actually I should check the perf cost of doing it every draw, it might not be _that_ bad with `if "numpy" in sys.modules:`...

**Comment by jobh** [comment]
> Unfortunately Hypothesis will then complain about `Flaky` errors when it goes to shrink such things.

Ah, yes, of course. We would have to flag the test and warn when done, I guess. No need to go that far in this PR IMO, it can be added later.

**Pull request merged**